src/mapport.cpp

```diff
@@ -0,0 +1,342 @@
+    const char * multicastif = nullptr;
+    const char * minissdpdpath = nullptr;
+    struct UPNPDev * devlist = nullptr;
+    char lanaddr[64];
+
+    int error = 0;
+#if MINIUPNPC_API_VERSION < 14
+    devlist = upnpDiscover(2000, multicastif, minissdpdpath, 0, 0, &error);
+#else
+    devlist = upnpDiscover(2000, multicastif, minissdpdpath, 0, 0, 2, &error);
+#endif
+
+    struct UPNPUrls urls;
+    struct IGDdatas data;
+    int r;
+
+    r = UPNP_GetValidIGD(devlist, &urls, &data, lanaddr, sizeof(lanaddr));
+    if (r == 1)
+    {
+        if (fDiscover) {
+            char externalIPAddress[40];
+            r = UPNP_GetExternalIPAddress(urls.controlURL, data.first.servicetype, externalIPAddress);
+            if (r != UPNPCOMMAND_SUCCESS) {
+                LogPrintf("UPnP: GetExternalIPAddress() returned %d\n", r);
+            } else {
+                if (externalIPAddress[0]) {
+                    CNetAddr resolved;
+                    if (LookupHost(externalIPAddress, resolved, false)) {
+                        LogPrintf("UPnP: ExternalIPAddress = %s\n", resolved.ToString());
+                        AddLocal(resolved, LOCAL_MAPPED);
+                    }
+                } else {
+                    LogPrintf("UPnP: GetExternalIPAddress failed.\n");
+                }
+            }
+        }
+
+        std::string strDesc = PACKAGE_NAME " " + FormatFullVersion();
+
+        do {
+            r = UPNP_AddPortMapping(urls.controlURL, data.first.servicetype, port.c_str(), port.c_str(), lanaddr, strDesc.c_str(), "TCP", 0, "0");
+
+            if (r != UPNPCOMMAND_SUCCESS) {
+                ret = false;
+                LogPrintf("AddPortMapping(%s, %s, %s) failed with code %d (%s)\n", port, port, lanaddr, r, strupnperror(r));
+                break;
+            } else {
+                ret = true;
+                LogPrintf("UPnP Port Mapping successful.\n");
+            }
+        } while (g_mapport_interrupt.sleep_for(PORT_MAPPING_REANNOUNCE_PERIOD));
+        g_mapport_interrupt.reset();
+
+        r = UPNP_DeletePortMapping(urls.controlURL, data.first.servicetype, port.c_str(), "TCP", 0);
+        LogPrintf("UPNP_DeletePortMapping() returned: %d\n", r);
+        freeUPNPDevlist(devlist); devlist = nullptr;
+        FreeUPNPUrls(&urls);
+    } else {
+        LogPrintf("No valid UPnP IGDs found\n");
+        freeUPNPDevlist(devlist); devlist = nullptr;
+        if (r != 0)
+            FreeUPNPUrls(&urls);
+    }
+
+    return ret;
+}
+#endif // #ifdef USE_UPNP
+
+static void ThreadMapPort()
+{
+    bool ok;
+    do {
+        ok = false;
+
+#ifdef USE_UPNP
+        // High priority protocol.
+        if (g_mapport_enabled_protos & MapPortProtoFlag::UPNP) {
+            g_mapport_current_proto = MapPortProtoFlag::UPNP;
+            ok = ProcessUpnp();
+            if (ok) continue;
+        }
+#endif // #ifdef USE_UPNP
+
+#ifdef USE_NATPMP
+        // Low priority protocol.
```

Comment:
Why is NATPMP "low priority"? I think NATPMP is the preferred protocol?

Reply:
https://github.com/bitcoin/bitcoin/pull/18077#issuecomment-588484002 by @luke-jr

Reply:
FWIW, I disagree. I think we should prefer NATPMP and phase out UPnP in the longer run when it has proven to be stable.

Reply:
@luke-jr, what do you mean by "maps random ports often"? Some implementation of natpmp has bugs? Can you give an example?

Maybe just remove this comment "// Low priority protocol." -- it is unclear what is meant by it and is missing in other places where `#ifdef USE_NATPMP` is used.

Reply:
@vasild 
> @luke-jr, what do you mean by "maps random ports often"? Some implementation of natpmp has bugs? Can you give an example?

Some NAT-PMP daemon implementations do randomize an external port. For example, see https://github.com/bitcoin/bitcoin/pull/18077#issuecomment-588121365 and http://bodgitandscarper.co.uk/natpmpd/:
> The external TCP or UDP port assigned to the client is always randomised rather than giving the first client the port it actually requested and then trying to work out what to do for additional clients that want the same external port.

Reply:
@hebasto, Thanks for the clarification! IMO the nat-pmp implementation overriding the desired external port is fine as long as we advertise the overriden one instead of the desired one (I guess we do). That should also be expected as it is allowed by the RFC.

I guess something like that happens with UPnP too if the desired port is already used?

Reply:
> IMO the nat-pmp implementation overriding the desired external port is fine as long as we advertise the overriden one instead of the desired one (I guess we do).

Yes, we do:
https://github.com/bitcoin/bitcoin/blob/890e59c64d62fd39778ede620a70fe6bc60d9993/src/mapport.cpp#L97-L102

> I guess something like that happens with UPnP too if the desired port is already used?

https://github.com/bitcoin/bitcoin/blob/890e59c64d62fd39778ede620a70fe6bc60d9993/src/mapport.cpp#L173-L186